Major refactoring (removed dependency on annotations, completed tests)

cleentfaar opened the pull request:
This removes the dependency on annotations for serialization of objects (now using yaml mapping).

I've also added almost all of the tests that were (still) missing, and fixed **a lot** of hidden bugs.

todo:
- [x] Stamp out as much bugs as possible
- [x] Add the (missing) tests for all payloads and responses
- [x] Add the missing tests for the models

deancsmith commented:
Check out my last couple of commits on my [master branch](https://github.com/deancsmith/slack/commits/master).

There appear to have been some recent changes to the postMessage/attachments structure:

https://api.slack.com/methods/chat.postMessage
https://api.slack.com/docs/attachments

I can add in the extra properties to the yaml (title, author*, etc.) once you merge in your above changes if that helps but I wasn't sure how you wanted to solve the below.

The thing I could figure out how to do nicely (see my third hacky commit) was to have the $payload->attachments property serialised as a JSON array (e.g. as a string) before it is actually sent to Slack. It looked like a HandlerCallback or a PreSerialization event was the way to go but I haven't really used the JMS/Serialization bundle much before.

It also looks like the $payload->attachments->fields needs to be converted to an ArrayCollection as well, as a field is no longer a key:value pair, but is now an array with three properties (title, value, short).

cleentfaar commented:
@deancsmith Thanks again for the feedback! :baby: 

I've refactored it a bit further. Have a look at the updated [Attachment](https://github.com/cleentfaar/slack/blob/removed-annotation-dependency-and-completed-tests/src/CL/Slack/Model/Attachment.php) and [AttachmentField](https://github.com/cleentfaar/slack/blob/removed-annotation-dependency-and-completed-tests/src/CL/Slack/Model/Attachment.php) models, I've also updated tests accordingly.

I've included all properties that are currently documented for the [Attachment](https://api.slack.com/docs/attachments) object you linked to.

I'm currently finalizing this PR so if you see anything else out of the ordinary let me know, cheers!

deancsmith commented:
Hey, great work on this! I still haven't had a chance to go back over my code and change my application to use attachments, rather than just a normal message with markdown as I did as a workaround.

Just looking at your code [here](https://github.com/cleentfaar/slack/blob/master/src/CL/Slack/Payload/ChatPostMessagePayload.php) I can't see how you got the attachments to be serialised as a string within the ChatPostMessage. From the documentation, and when I was playing around, it looked like the attachments property needed to be a string within the ChatPostMessage array rather than an array itself. If that makes sense? Did you test this against slack?

If not, I hope to have time to revisit this later in the week.
